aten/src/ATen/native/nested/NestedTensorTransformerFunctions.cpp

```diff
@@ -260,5 +261,128 @@ Tensor NestedTensor_to_mask(const Tensor& nt, c10::optional<int64_t> mask_dim, c
+  TORCH_CHECK(
+      qkv.is_nested(),
+      "QKV must be nested for flash cumulative_seq_len calculation.")
+  auto* nt_impl = get_nested_tensor_impl(qkv);
+  const auto& sizes = nt_impl->get_nested_size_tensor();
+  auto size_tensor_stride = sizes.stride(0);
+
+  const int64_t batch_size = qkv.size(0);
+  auto cumulative_seqlen = at::zeros({batch_size + 1}, TensorOptions().device(at::kCPU).dtype(at::kInt));
+
+  auto* sizes_ptr = sizes.data_ptr<int64_t>();
+  auto* cumulative_seqlen_ptr = cumulative_seqlen.data_ptr<int32_t>();
+
+  int32_t sum = 0;
+  int64_t max_seqlen = -1;
+  cumulative_seqlen_ptr[0] = sum;
+  for (const auto i : c10::irange(batch_size)) {
+    // Calculate the cumulative sum of the sequence lengths
+    auto current_seq_len = sizes_ptr[i * size_tensor_stride];
+    sum += current_seq_len;
+    cumulative_seqlen_ptr[i + 1] = sum;
+
+    // Find the max element while we traverse
+    max_seqlen = std::max(max_seqlen, current_seq_len);
+  }
+  // Send to GPU, this is pretty light weight calc for normal batch size
+  // but maybe this needs to be on gpu
+  cumulative_seqlen = cumulative_seqlen.to(TensorOptions().device(at::kCUDA));
+  return std::tuple<Tensor, int64_t>{cumulative_seqlen, max_seqlen};
+}
+
+struct {
+  Tensor q;
+  Tensor k;
+  Tensor v;
+  Tensor cumulative_sequence_length_q;
+  Tensor cumulative_sequence_length_k;
+  int64_t max_seqlen_q{};
+  int64_t max_seqlen_k{};
+} flash_attention_data;
+
+Tensor flash_attention_helper(
+    const Tensor& query,
+    const Tensor& key,
+    const Tensor& value,
+    int64_t num_heads,
+    double dropout_p,
+    bool causal) {
+  //  Query is of size (batch_size x ragged_seq_len x (potentially 3* n_heads *
+  //  head_dim))
+  auto cumulative_and_max_q = cumulative_and_max_seq_len(query);
+  Tensor cumulative_sequence_length_q = std::get<0>(cumulative_and_max_q);
+  int64_t max_seqlen_batch_q = std::get<1>(cumulative_and_max_q);
+
+  if (key.is_same(value) || query.is_same(key) || query.is_same(value)) {
+    int64_t head_dim{query.size(-1) / (num_heads * 3)};
+    int64_t Nnz_q{cumulative_sequence_length_q[-1].item<int64_t>()};
+
+    // The input/output of flash attention is (Nnz, nheads, headdim) we need to
+    // create a nested size with the last dim = nheads*headdim If q,k,v are all
+    // equal than the current nt last dim is of size nheads*headdim*3 after
+    // projection
+    auto atten_size = get_nested_size_tensor(query);
+    auto div3 = atten_size.index({at::indexing::Slice(), -1}).div(3);
+    atten_size.index({at::indexing::Slice(), -1}) = div3;
+
+    auto qkv_buffer_reshaped =
+        get_buffer(query).view({Nnz_q, 3, num_heads, head_dim});
+
+    // If we are passing in query, key, value all the same tensors than we have
+    // packed them into one tensor and need to slice for flash attention
+    Tensor atten_buffer = at::_flash_scaled_dot_product_attention(
+        qkv_buffer_reshaped.index({at::indexing::Slice(), 0}),
```

Comment:
Require users to do the slicing beforehand. I don't see `_flash_scaled_dot_product_attention` requiring that they are all the same. Do you even need this helper or can you call into `_flash_scaled_dot_product_attention` and then do the `cumulative_sequence_length_q` and `max_seqlen_batch_q` within it?

Reply:
I was getting a `wrap_buffer` undefined in the libcuda.so and for my sanity and just to help implement I broke these up. But since this is only a cuda kernel it should probably all live close to one place. Yeah slicing is probably the way to go. Currently slicing can't be done on nested tensors. I think some of Yifens reshape work will help here but that is stalled up by some internal error.